dsumaroka/jv-fruit-shop

dsumaroka opened the pull request.

olekskov commented on src/main/java/core/basesyntax/Main.java:
```diff
@@ -0,0 +1,11 @@
+
+public class Main {
+    public static void main(String[] args) {
+        FruitProcessing fruitProcessing = new FruitProcessingImpl();
+        fruitProcessing.fruitProcessing("inputData.csv");
```
Now you have single service that doing enormous amount of tasks and for unprepared person it's vary hard to understand the flow.

In you main method you should do following things:
1. Some new service (e.g. FileReader) will read content of input file (probably to List<String>)
2. Some other service (e.g. TransactionParser) will parse strings to some mode (e.g. FruitTransaction)
3. One more service will take these transations and change state of some DB analog (Map<String, Integer> in separate class Storage is easiest option to go with, where key is fruit name and value is quantity) processing them one-by-one.
4. Service from point 3 will require strategy, which will provide proper TransactionHandler depending on Transaction type.
5. And handlers will contain method to change storage state (each for operation it represents)
6. After all of that there will be additional service that generates report text based on data in storage.
7. And then some FileWriter service will take this report and write it to output file

olekskov commented on src/main/java/core/basesyntax/service/impl/FruitProcessingImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public void fruitProcessing(String fileName) {
+        FruitDao fruitDao = new FruitDaoImpl();
+        List<String> fruitsForProcessing = fruitDao.getInputData(fileName);
+        StringBuilder report = new StringBuilder("fruit,quantity" + System.lineSeparator());
+        while (fruitsForProcessing.size() != 1) {
+            String fruitName = Arrays
+                    .stream(fruitsForProcessing.get(1).trim().split(DELIMITER_REGEX))
+                    .skip(1)
+                    .limit(1)
+                    .collect(Collectors.joining());
+            int balance = 0;
+            List<String> oneTypeFruit = fruitsForProcessing.stream()
+                    .filter(fr -> fr.contains(fruitName.trim()))
+                    .toList();
+            for (String fruit : oneTypeFruit) {
+                String[] fruitTransQuantData = fruit.split(DELIMITER_REGEX);
+                balance = new FruitTransaction()
+                        .getStrategy(fruitTransQuantData[FRUIT_OPERATION_TYPE_INDEX].trim())
+                        .balanceUpdater(balance, fruitTransQuantData[FRUIT_QUANTITY_INDEX].trim());
+            }
+            fruitsForProcessing.removeAll(oneTypeFruit);
+            report.append(fruitName).append(",")
+                    .append(balance)
+                    .append(System.lineSeparator());
+        }
+        fruitDao.sendResultData(report.toString());
```
This should be redesigned according to comment above. You need much bigger amount of services to make code understandable and maintainable

olekskov commented on src/main/java/core/basesyntax/strategy/TransactionStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.strategy;
+
+public interface TransactionStrategy {
+    int balanceUpdater(int balance, String quantity);
+}
```
Task of strategy pattern is to provide proper implementation of some interface depending on input
so it should be something like
```suggestion
public interface TransactionStrategy {
    TransactionHandler getHandler(FruitTransaction transaction);
}
```

dsumaroka commented:
Ok

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Никаких
вирусов.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

сб, 10 февр. 2024 г. в 18:38, Oleksandr Kovalenko ***@***.***
>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/main/java/core/basesyntax/Main.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485192663>
> :
>
> > +        FruitProcessing fruitProcessing = new FruitProcessingImpl();
> +        fruitProcessing.fruitProcessing("inputData.csv");
>
> Now you have single service that doing enormous amount of tasks and for
> unprepared person it's vary hard to understand the flow.
>
> In you main method you should do following things:
>
>    1. Some new service (e.g. FileReader) will read content of input file
>    (probably to List)
>    2. Some other service (e.g. TransactionParser) will parse strings to
>    some mode (e.g. FruitTransaction)
>    3. One more service will take these transations and change state of
>    some DB analog (Map<String, Integer> in separate class Storage is easiest
>    option to go with, where key is fruit name and value is quantity)
>    processing them one-by-one.
>    4. Service from point 3 will require strategy, which will provide
>    proper TransactionHandler depending on Transaction type.
>    5. And handlers will contain method to change storage state (each for
>    operation it represents)
>    6. After all of that there will be additional service that generates
>    report text based on data in storage.
>    7. And then some FileWriter service will take this report and write it
>    to output file
>
> ------------------------------
>
> In src/main/java/core/basesyntax/service/impl/FruitProcessingImpl.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485193056>
> :
>
> > +        FruitDao fruitDao = new FruitDaoImpl();
> +        List<String> fruitsForProcessing = fruitDao.getInputData(fileName);
> +        StringBuilder report = new StringBuilder("fruit,quantity" + System.lineSeparator());
> +        while (fruitsForProcessing.size() != 1) {
> +            String fruitName = Arrays
> +                    .stream(fruitsForProcessing.get(1).trim().split(DELIMITER_REGEX))
> +                    .skip(1)
> +                    .limit(1)
> +                    .collect(Collectors.joining());
> +            int balance = 0;
> +            List<String> oneTypeFruit = fruitsForProcessing.stream()
> +                    .filter(fr -> fr.contains(fruitName.trim()))
> +                    .toList();
> +            for (String fruit : oneTypeFruit) {
> +                String[] fruitTransQuantData = fruit.split(DELIMITER_REGEX);
> +                balance = new FruitTransaction()
> +                        .getStrategy(fruitTransQuantData[FRUIT_OPERATION_TYPE_INDEX].trim())
> +                        .balanceUpdater(balance, fruitTransQuantData[FRUIT_QUANTITY_INDEX].trim());
> +            }
> +            fruitsForProcessing.removeAll(oneTypeFruit);
> +            report.append(fruitName).append(",")
> +                    .append(balance)
> +                    .append(System.lineSeparator());
> +        }
> +        fruitDao.sendResultData(report.toString());
>
> This should be redesigned according to comment above. You need much bigger
> amount of services to make code understandable and maintainable
> ------------------------------
>
> In src/main/java/core/basesyntax/strategy/TransactionStrategy.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485193577>
> :
>
> > +public interface TransactionStrategy {
> +    int balanceUpdater(int balance, String quantity);
> +}
>
> Task of strategy pattern is to provide proper implementation of some
> interface depending on input
> so it should be something like
> ⬇️ Suggested change
>
> -public interface TransactionStrategy {
> -    int balanceUpdater(int balance, String quantity);
> -}
> +public interface TransactionStrategy {
> +    TransactionHandler getHandler(FruitTransaction transaction);
> +}
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#pullrequestreview-1873738063>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7UEH4FHBXGH2ZL7TFD2VU3YS6PB3AVCNFSM6AAAAABDCXN6LSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZTG4ZTQMBWGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Vasul-Pankiv commented on src/main/java/core/basesyntax/strategy/TransactionStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.strategy;
+
+public interface TransactionStrategy {
+    Integer balanceUpdater(String fruitName, int quantity);
```
You have to refactor your strategy, do not ignore comments
![image](https://github.com/mate-academy/jv-fruit-shop/assets/48316721/40892b68-83e5-4cc6-9bdf-8e8204284fd0)

Vasul-Pankiv commented on src/main/java/core/basesyntax/strategy/TransactionStrategyImpl.java:
```diff
@@ -0,0 +1,17 @@
+public class TransactionStrategyImpl implements TransactionStrategy {
+    @Override
+    public Integer balanceUpdater(String fruitName, int quantity) {
+        Storage storage = new Storage();
```
You should have only one object of storage for all your app. Also a greate idea is to add DAO layer to provide access to the storage

Vasul-Pankiv commented on src/main/java/core/basesyntax/service/TransactionParser.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.List;
+
+public interface TransactionParser {
+    List<FruitTransaction> parsOf(List<String> fruits);
```
```suggestion
    List<FruitTransaction> parse(List<String> fruits);
```

Vasul-Pankiv commented on src/main/java/core/basesyntax/service/impl/FileReaderImpl.java:
```diff
@@ -0,0 +1,25 @@
+    public List<String> getDataFromInputFile(String fileName) {
+        File inputDataFile = new File("src/main/resources/" + fileName);
+        try {
+            inputDataFile.createNewFile();
```
why are you creating file here? File reader has to only read existing file

kozhukhovsky replied:
not fixed

Vasul-Pankiv commented on src/main/java/core/basesyntax/service/FruitProcessing.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.List;
+import java.util.Map;
+
+public interface FruitProcessing {
+    Map<String, Integer> fruitProcessing(List<FruitTransaction> fruitTransactions);
```
```suggestion
public interface TransactionExecutor {
    Map<String, Integer> executeAll(List<FruitTransaction> fruitTransactions);
```

dsumaroka commented:
Ок

Пн, 12 февр. 2024 г. в 12:10, Vasyl Pankiv ***@***.***>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/main/java/core/basesyntax/db/Storage.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485945682>
> :
>
> > @@ -0,0 +1,12 @@
> +package core.basesyntax.db;
> +
> +import java.util.HashMap;
> +import java.util.Map;
> +
> +public class Storage {
>
> Storage has to contain data about fruits, not about trnsactions
> ------------------------------
>
> In src/main/java/core/basesyntax/service/FileReader.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485947550>
> :
>
> > @@ -0,0 +1,7 @@
> +package core.basesyntax.service;
> +
> +import java.util.List;
> +
> +public interface FileReader {
> +    List<String> getDataFromInputFile(String fileName);
>
> ⬇️ Suggested change
>
> -    List<String> getDataFromInputFile(String fileName);
> +    List<String> readFile(String fileName);
>
> ------------------------------
>
> In src/main/java/core/basesyntax/service/FileWriter.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485948350>
> :
>
> > @@ -0,0 +1,5 @@
> +package core.basesyntax.service;
> +
> +public interface FileWriter {
> +    void sendResultData(String report);
>
> ⬇️ Suggested change
>
> -    void sendResultData(String report);
> +    void writeToFile(String report);
>
> ------------------------------
>
> In src/main/java/core/basesyntax/strategy/TransactionStrategy.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485954498>
> :
>
> > @@ -0,0 +1,5 @@
> +package core.basesyntax.strategy;
> +
> +public interface TransactionStrategy {
> +    Integer balanceUpdater(String fruitName, int quantity);
>
> You have to refactor your strategy, do not ignore comments
> image.png (view on web)
> <https://github.com/mate-academy/jv-fruit-shop/assets/48316721/40892b68-83e5-4cc6-9bdf-8e8204284fd0>
> ------------------------------
>
> In src/main/java/core/basesyntax/strategy/TransactionStrategyImpl.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485956326>
> :
>
> > @@ -0,0 +1,17 @@
> +package core.basesyntax.strategy;
> +
> +import core.basesyntax.db.Storage;
> +
> +public class TransactionStrategyImpl implements TransactionStrategy {
> +    @Override
> +    public Integer balanceUpdater(String fruitName, int quantity) {
> +        Storage storage = new Storage();
>
> You should have only one object of storage for all your app. Also a greate
> idea is to add DAO layer to provide access to the storage
> ------------------------------
>
> In src/main/java/core/basesyntax/service/TransactionParser.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485960168>
> :
>
> > @@ -0,0 +1,8 @@
> +package core.basesyntax.service;
> +
> +import core.basesyntax.model.FruitTransaction;
> +import java.util.List;
> +
> +public interface TransactionParser {
> +    List<FruitTransaction> parsOf(List<String> fruits);
>
> ⬇️ Suggested change
>
> -    List<FruitTransaction> parsOf(List<String> fruits);
> +    List<FruitTransaction> parse(List<String> fruits);
>
> ------------------------------
>
> In src/main/java/core/basesyntax/service/impl/FileReaderImpl.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485961852>
> :
>
> > @@ -0,0 +1,25 @@
> +package core.basesyntax.service.impl;
> +
> +import core.basesyntax.service.FileReader;
> +import java.io.File;
> +import java.io.IOException;
> +import java.nio.file.Files;
> +import java.util.List;
> +
> +public class FileReaderImpl implements FileReader {
> +
> +    @Override
> +    public List<String> getDataFromInputFile(String fileName) {
> +        File inputDataFile = new File("src/main/resources/" + fileName);
> +        try {
> +            inputDataFile.createNewFile();
>
> why are you creating file here? File reader has to only read existing file
> ------------------------------
>
> In src/main/java/core/basesyntax/service/FruitProcessing.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1485965416>
> :
>
> > +public interface FruitProcessing {
> +    Map<String, Integer> fruitProcessing(List<FruitTransaction> fruitTransactions);
>
> ⬇️ Suggested change
>
> -public interface FruitProcessing {
> -    Map<String, Integer> fruitProcessing(List<FruitTransaction> fruitTransactions);
> +public interface TransactionExecutor {
> +    Map<String, Integer> executeAll(List<FruitTransaction> fruitTransactions);
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#pullrequestreview-1874861234>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7UEH4HQYHEWTJ7PYATOFT3YTHTA7AVCNFSM6AAAAABDCXN6LSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZUHA3DCMRTGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dsumaroka commented:
Sorry, I fixed it locally, but got confused in the commits and sent the
wrong version. I'll fix it

ср, 14 февр. 2024 г. в 14:19, Anton Kozhukhivskiy ***@***.***
>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/core/basesyntax/service/impl/FileReaderImpl.java
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1489382064>
> :
>
> > @@ -0,0 +1,25 @@
> +package core.basesyntax.service.impl;
> +
> +import core.basesyntax.service.FileReader;
> +import java.io.File;
> +import java.io.IOException;
> +import java.nio.file.Files;
> +import java.util.List;
> +
> +public class FileReaderImpl implements FileReader {
> +
> +    @Override
> +    public List<String> getDataFromInputFile(String fileName) {
> +        File inputDataFile = new File("src/main/resources/" + fileName);
> +        try {
> +            inputDataFile.createNewFile();
>
> not fixed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/jv-fruit-shop/pull/1058#discussion_r1489382064>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7UEH4ALFBGSTEBU2ND7NK3YTSTUDAVCNFSM6AAAAABDCXN6LSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQOBQGEZTQMBUGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
